src/FSharp.Core/prim-types.fs

```diff
@@ -5206,6 +5217,29 @@ namespace Microsoft.FSharp.Core
                 if value = unsafeDefault<'T> then ""
                 else let x = (# "" value : IFormattable #) in defaultIfNull "" (x.ToString(null, CultureInfo.InvariantCulture))
 
+        [<CompiledName("ToString")>]
+        let inline string (value: 'T) = 
+             defaultString value
+
+             // Only compilers that understand `when 'T : SupportsWhenTEnum` will understand `when 'T : Enum`.
+             when 'T : SupportsWhenTEnum =
```

Comment:
This is a new mechanism for making language feature toggles which are strongly coupled between fsharp.core and the compiler.

When I was first reading the code (before reading your PR's description), I thought that it related to T itself.
I would want to come up with a naming scheme that makes it apparent that this is indeed a feature toggle.

What about a dedicated module/namespace to host these marker types?
like `'T : CompilerLibraryFeatures.SupportsWhenTEnum` or something indicating that naming?


= I want it to be apparent that it decides based on compiler capabilities

Reply:
Or maybe `CompilerServices.SupportsWhenTEnumFeature`

Reply:
Is `when 'T : CompilerServices.SupportsWhenTEnum` OK, or would you prefer the `Feature` suffix? Or there could even be a `Features` module or something, although I don't expect we'll be adding many (if any) more of these.